### PR TITLE
[fix] tensor_split : non-contiguous indices tensor

### LINKED ISSUE #63281: tensor_split does not handle non-contiguous indices
## 🐛 Bug

torch.tensor_split does not produce the same outputs for non-contiguous tensors as for contiguous, see repro below.

## To Reproduce

Steps to reproduce the bug:

```python
inner_indices = torch.tensor([[564,   0],
        [564,   0],
        [564,   0],
        [564,   1],
        [565,   1],
        [566,   0],
        [566,   0],
        [566,   0],
        [566,   0],
        [566,   0],
        [566,   0],
        [566,   1],
        [567,   0],
        [567,   0],
        [567,   0],
        [567,   0],
        [567,   0],
        [567,   1],
        [568,   0],
        [568,   0],
        [568,   0],
        [568,   1],
        [569,   0],
        [569,   0],
        [569,   0],
        [569,   1],
        [570,   0],
        [570,   0],
        [570,   0],
        [570,   0],
        [570,   0]])
data = torch.tensor([4, 4, 4, 4, 4, 4], dtype=torch.int32)
for u, v in zip(torch.tensor_split(data, inner_indices[:, 1]),
                torch.tensor_split(data, inner_indices[:, 1].contiguous())):
    if u.shape != v.shape:
        print(u, v)
```

Output:
```
tensor([4, 4, 4, 4, 4, 4], dtype=torch.int32) tensor([], dtype=torch.int32)
tensor([4, 4, 4, 4, 4, 4], dtype=torch.int32) tensor([4], dtype=torch.int32)
tensor([4, 4, 4, 4, 4, 4], dtype=torch.int32) tensor([], dtype=torch.int32)
tensor([4, 4, 4, 4, 4], dtype=torch.int32) tensor([], dtype=torch.int32)
tensor([4, 4, 4, 4, 4], dtype=torch.int32) tensor([], dtype=torch.int32)
tensor([4, 4, 4, 4, 4, 4], dtype=torch.int32) tensor([4], dtype=torch.int32)
tensor([4, 4, 4, 4, 4, 4], dtype=torch.int32) tensor([], dtype=torch.int32)
tensor([4, 4, 4, 4, 4, 4], dtype=torch.int32) tensor([], dtype=torch.int32)
tensor([4, 4, 4, 4, 4, 4], dtype=torch.int32) tensor([4], dtype=torch.int32)
tensor([4, 4, 4, 4, 4, 4], dtype=torch.int32) tensor([], dtype=torch.int32)
tensor([4, 4, 4, 4, 4, 4], dtype=torch.int32) tensor([4], dtype=torch.int32)
tensor([4, 4, 4, 4, 4], dtype=torch.int32) tensor([], dtype=torch.int32)
tensor([4, 4, 4, 4, 4, 4], dtype=torch.int32) tensor([4], dtype=torch.int32)
tensor([4, 4, 4, 4, 4, 4], dtype=torch.int32) tensor([], dtype=torch.int32)
tensor([4, 4, 4, 4, 4, 4], dtype=torch.int32) tensor([], dtype=torch.int32)
```

## Expected behavior

Output should be the same for contiguous and non-contiguous tensors.

## Environment

Tested in stable 1.9 and nightly '1.10.0.dev20210814+cpu'.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @VitalyFedyunin @jamesr66a @ppwwyyxx

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes https://github.com/pytorch/pytorch/issues/63281

### CODE REVIEW

**Comment by kshitij12345** [comment]
Should we consolidate the reference_test in `test_view_ops.py` to OpInfo (maybe in future PR)

https://github.com/pytorch/pytorch/blob/22f78144c70d9d215bc390805aa8d0657e548b85/test/test_view_ops.py#L1383-L1412

**Comment by mruberry** [comment]
> Should we consolidate the reference_test in `test_view_ops.py` to OpInfo (maybe in future PR)
> 
> https://github.com/pytorch/pytorch/blob/22f78144c70d9d215bc390805aa8d0657e548b85/test/test_view_ops.py#L1383-L1412

Maybe... but in general I think the tests in test_view_ops.py would be hard to automate

**Review by mruberry — APPROVED** [review]
Cool!

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -609,7 +609,13 @@ std::vector<Tensor> tensor_split(const Tensor& self, const Tensor& tensor_indice
     return self.tensor_split(sections, dim);
   } else {
     auto indices_data = tensor_indices_or_sections.data_ptr<int64_t>();
-    std::vector<int64_t> indices(indices_data, indices_data + tensor_indices_or_sections.numel());
+    auto stride = tensor_indices_or_sections.stride(0);
+    auto numel = tensor_indices_or_sections.numel();
+    std::vector<int64_t> indices(numel);
+    for (size_t offset = 0; offset < numel; offset++) {
```
Comment (mruberry) [inline aten/src/ATen/native/TensorShape.cpp:615]: C++ style nit is that offset should be typed like numel, but it's fine

**Comment by mruberry** [comment]
Did you manually run the exact scenario described in the issue, too?

**Comment by kshitij12345** [comment]
> Did you manually run the exact scenario described in the issue, too?

Yes
image

**Pull request closed without merging**